utils/siprealm.1
.\" siprealm - show or set authentication realm of sipwitch services.
.\" Copyright (c) 2010 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH siprealm "1" "January 2010" "GNU SIP Witch" "GNU Telephony"
.SH NAME
siprealm \- show or set authentication realm of sipwitch services
.SH SYNOPSIS
.B siprealm [ 
.I [-md5|-sha1|-rmd160] 
.B ''realm'']
.br
.SH DESCRIPTION
\fBSiprealm\fR is used to either set or display the current authentication
realm of sipwitch services.  This is used to authenticate sip clients.  If
you are aggegating a group of sipwitch servers from multiple hosts, they should
all share the same realm.  If you are running NIS services then it is a
reasonable practice to set the realm of your machines to match your NIS domain.
If you are using a stand-alone sipwitch, such as for an independent desktop
workstation, you might want to use a uuid.  
.PP
When a proper domain is used as the realm, then sipwitch behaves as if
the old stack domain property is also set.  Realms that look like proper
domain names are required to perform internode networking, since the
identity presented between nodes must be common (based on the domain) and
the authentication realm must also match each node for digests to work between
nodes.  Hence, it became much simpler and more consistent to combine 
authentication realms and identity domains.
.PP
Only the root user can set the sipwitch domain or hash type.  When either is
changed, any sipwitch digest entries cached are automatically cleared.  This
means that if you have sipwitch setup to generate digests based on user
passwords set through "pam", then you will have to reset or update each user's
password if you change the realm.  For this reason, it's generally advised to
set it only once. 
.SH OPTIONS
.TP
.B -md5
Use md5 digest algorithm (default).
.TP
.B -rmd160
Use rmd160 digest algorithm.
.TP
.B -sha1
Use sha1 digest algorithm.
.SH "EXIT STATUS"
Any error in argument format will return an exit status of 3.  If no realm is
set and used to print the current realm, then an exist status of 1 will be
returned.  The command will normally return with exit status of 0 after 
printing the curret (or new) sip realm.
.SH AUTHOR
.B siprealm
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to sipwitch-devel@gnu.org.
.SH COPYRIGHT
Copyright \(co 2010 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



NEWS
SIP Witch NEWS -- history of visible user changes. Updated 2008-04-15

SIP Witch is an official package of the GNU Project as of August 10th 2007.
GNU SIP Witch is also part of GNU Telephony & the GNU Telecom subsystem.

sipwitch 0.1 codebase initial public release with support of SIP registration,
multi-target registration, authentication peering over SIP, and registration
querying of user agents, basic call processing between registered user agents
including call distribution for multi-target registrations, and basic SIP
instant messaging support.

sipwitch 0.4 represents the first minimally deployable baseline sipwitch
server.  This release includes multipath calling, and basic call forwarding.
Also important with this release is stablization of core management api's,
including SWIG interfaces for local scripting languages, and xmlrpc cgi.
Basic management and administration services can now be built for sipwitch.
Support for changing server state is also complete, such as for "day"/"night"
modes, holidays, etc.  Basic server call statistics and call detail record
processing is completed.

sipwitch 0.5 introduces the concept of internodal calling and allows sipwitch
to transform itself into a sip redirection server for certain local call
destinations.  Internodal calling differs from dialing to remote uri's in that
the authentication record is pre-stuffed into the remote invite, hence, all
nodes must share the same sip realm.  This allows one sipwitch server to call a
"local" extension number that actually appears on a remote switch rather than
the one the user agent is registered to. Internodal references are generated
using pattern rules through the static <routing> table <redirect> entries.  The
forward secure calling domain plugin also makes use of internodal calling to
connect with the insecure domain.
  


utils/siprealm.cpp
// Copyright (C) 2010 David Sugar, Tycho Softworks.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/ucommon.h>
#include <config.h>

using namespace UCOMMON_NAMESPACE;

static void usage(void)
{
	fprintf(stderr, "usage: siprealm [[-md5|-rmd160|-sha1] [\"realm\"]]\n"
		"[options]\n"
		"  -md5                 specify md5 computation\n"
		"  -rmd160              specify rmd160 computation\n"
		"  -sha1                specify sha1 computation\n"
		"Report bugs to sipwitch-devel@gnu.org\n");
	exit(3);
}

static void show(void)
{
	fsys_t fs;
	char buffer[256];

	fsys::open(fs, DEFAULT_CFGPATH "/siprealm", fsys::ACCESS_RDONLY);
	if(!is(fs))
		fsys::open(fs, DEFAULT_VARPATH "/lib/sipwitch/uuid", fsys::ACCESS_RDONLY);

	if(!is(fs)) {
error:
		fprintf(stderr, "*** siprealm: no public realm known\n");
        exit(1);
    }

	memset(buffer, 0, sizeof(buffer));
	fsys::read(fs, buffer, sizeof(buffer) - 1);
	fsys::close(fs);

	char *cp = strchr(buffer, '\n');
	if(cp)
		*cp = 0;

	cp = strchr(buffer, ':');
	if(cp)
		*cp = 0;

	if(!buffer[0])
		goto error;

	printf("%s\n", buffer);
	exit(0);
}

extern "C" int main(int argc, char **argv)
{
	char *realm = NULL;
	const char *mode = NULL;
	fsys_t fs;
	char buffer[256];
	char replace[256];
	char *cp = NULL;
	FILE *fp;

#ifdef	_MSWINDOWS_
	const char *control = "\\\\.\\mailslot\\sipwitch_ctrl";
#else
	const char *control = DEFAULT_VARPATH "/run/sipwitch/control";
#endif

	++argv;
	if(!*argv)
		show();

	if(!strncmp(*argv, "--", 2))
		++*argv;

	if(!strcmp(*argv, "-version")) {
		printf("siprealm 0.1\n"
			"Copyright (C) 2008 David Sugar, Tycho Softworks\n"
			"License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
			"This is free software: you are free to change and redistribute it.\n"
			"There is NO WARRANTY, to the extent permitted by law.\n");
		exit(0);
	}

	if(String::equal(*argv, "-md5")) {
		mode = "MD5";
		++argv;
	}
	else if(String::equal(*argv, "-sha1")) {
		mode = "SHA1";
		++argv;
	}
	else if(String::equal(*argv, "-rmd160")) {
		mode = "RMD160";
		++argv;
	}
	else if(String::equal(*argv, "-?") || String::equal(*argv, "-help"))
		usage();
	else if(**argv == '-') {
		fprintf(stderr, "*** %s: unknown digest mode", *argv);
		exit(2);
	}

	fsys::open(fs, DEFAULT_CFGPATH "/siprealm", fsys::ACCESS_RDONLY);
	memset(buffer, 0, sizeof(buffer));
	if(is(fs)) {
		fsys::read(fs, buffer, sizeof(buffer) - 1);
		fsys::close(fs);
		cp = strchr(buffer, ':');
		if(cp)
			*(cp++) = 0;
	}

	realm = *argv;
	if(!realm)
		realm = buffer;
	else if(argv[1] || *realm == '-')
		usage();

	if(!cp || !*cp)
		cp = (char *)"md5";

	// make sure we have a valid mode, default is md5...
	if(!mode && cp && *cp)
		mode = cp;

	if(!mode)
		mode = "md5";

	// if unchanged, we leave alone...
	if(String::equal(buffer, realm) && String::equal(cp, mode))
		goto exit;

	// create replacement realm string...
	if(String::equal(mode, "md5"))
		String::set(replace, sizeof(replace), realm);
	else
		snprintf(replace, sizeof(replace), "%s:%s", realm, mode);

	::remove(DEFAULT_CFGPATH "/siprealm");
	fsys::create(fs, DEFAULT_CFGPATH "/siprealm", fsys::ACCESS_WRONLY, 0644);
	if(is(fs)) {
		fsys::write(fs, replace, strlen(replace));
		fsys::close(fs);
	}
	else {
		fprintf(stderr, "*** siprealm: root permission required\n");
		exit(0);
	}				

	// if previous digests cached, clear them as they are now invalid...
	::remove(DEFAULT_VARPATH "/lib/sipwitch/digests.db");

	// if server is up, also sync server with realm change...
	fp = fopen(control, "w");
	if(fp) {
		fprintf(fp, "realm %s\n", realm);
		fclose(fp);
	}

exit:
	printf("%s\n", realm);
	exit(0);
}

